keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,776 @@
+
+#### Provision discontiguous ranges
+A user wants to create a cluster with 32 nodes each with a capacity to run 256 pods. This means that each node needs a /24 pod CIDR range and they need a total range of /19. However, there aren't enough contiguous IPs in the user's network. They can find 4 free ranges of size /21 but no single contiguous /19 range.
+
+Using this enhancement, the user creates 4 different CIDR configurations each with a /21 range. The CIDR allocator allocates a /24 range from any of these /21 ranges to the nodes and the user can now create the cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+A major precondition for this feature is to prevent churn on the IPs assigned to Nodes. Specifically, we want to prevent the controller from attempting to change a Node’s assigned PodCIDRs. For example, K8s will reject attempts to delete a CIDRRange that’s currently in use by a Node. This should prevent accidental broad-ranging changes to the cluster.
+
+### Risks and Mitigations
+
+- Do racing controllers need a more sophisticated design? The current plan is to plug in to the kube-controller-manager leader election.
+- Can the controller install admission controls? This is a new pattern for in-tree components.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New Resource
+```
+ClusterCIDRConfig {
+    # Nodes may only have 1 range from each family.
+    # TODO: maybe this is redundant and can just be inferred by the controller.
+    IPFamily        FamilyType
+    
+    # An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    # +required
+    IPCIDRBlock     string
+    
+    # This defines which nodes the config is applicable to. An empty selector
+    # can be applied to any node.
+    # +optional
+    NodeSelector    v1.LabelSelector
+
+    # Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    # Users would have to ensure that the kubelet doesn't try to schedule 
+    # more pods than are supported by the node's netmask (i.e. the kubelet's
+    # --max-pods flag)
+    # +required
+    PerNodeMaskSize int
+}
+
+var (
+    IPV4 FamilyType = "IPV4"
+    IPV6 FamilyType = "IPV6"
+)
+```
+
+#### Expected Behavior
+
+- Each node will be assigned up to one range from each `FamilyType`.
+    - **TODO:** what is the tie-break if multiple ranges match
+- An empty `NodeSelector` functions as a default that applies to all nodes. This should be the fall-back and not take precedence if any other range matches.
```

Comment:
This is something we hadn't thought of. In this particular case, the end goal is to select only one range. We cannot match all selectors if they are across range objects. If we have two CIDR range objects range1 with node selector "color:blue" and range2 with node selector "size:small", it is unclear what range is being selected for a small blue node.

Maybe we say that the behaviour is undefined. Or maybe we add a new label type 'RangeSelector' and force all nodes to carry a label of that type?

Reply:
I was looking at a few other examples and I think the nodeSelector picking a random appropriate range is not unprecedented.
1. Pod's nodeSelector -- In this case the pod will be scheduled on any node that matches the selector. The semantics of pods make it obvious that we would only schedule it once. Similarly, my argument here is that the semantics of a node's PodCIDRs field make it "obvious" that we allocate only a single range from each IP Family
2. Persistent Volumes and Claims (might have misunderstood this) -- The documentation [here](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#selector) suggests that we filter the PVs applicable to a claim down to only PVs that match the selector and then pick any one which satisfies the claims other requirements

Based  on those examples, I'm thinking that our behavior be as follows:

> When attempting to allocate PodCIDRs for a Node, filter the ClusterCIDRConfigs down to those whose  NodeSelector matches that node. If one of the ClusterCIDRConfigs still has free IPs, allocate from one arbitrarily. Otherwise fall-back to ClusterCIDRConfigs with no selector.

What do you think @danwinship?